src/py/flwr/client/middleware/secure_aggregation/secaggplus_middleware.py

```diff
@@ -112,70 +113,115 @@ class SecAggPlusState:
     public_keys_dict: Dict[int, Tuple[bytes, bytes]] = field(default_factory=dict)
 
-    client: Optional[Union[Client, NumPyClient]] = None
-
-
-class SecAggPlusHandler(SecureAggregationHandler):
-    """Message handler for the SecAgg+ protocol."""
-
-    _shared_state = SecAggPlusState()
-    _current_stage = STAGE_UNMASK
-
-    def handle_secure_aggregation(
-        self, named_values: Dict[str, Value]
-    ) -> Dict[str, Value]:
-        """Handle incoming message and return results, following the SecAgg+ protocol.
-
-        Parameters
-        ----------
-        named_values : Dict[str, Value]
-            The named values retrieved from the SecureAggregation sub-message
-            of Task message in the server's TaskIns.
+    def __init__(self, **kwargs: ConfigsRecordValues) -> None:
+        for k, v in kwargs.items():
+            if k.endswith(":V"):
+                continue
+            new_v: Any = v
+            if k.endswith(":K"):
+                k = k[:-2]
+                keys = cast(List[int], v)
+                values = cast(List[bytes], kwargs[f"{k}:V"])
+                if len(values) > len(keys):
+                    updated_values = [
+                        tuple(values[i : i + 2]) for i in range(0, len(values), 2)
+                    ]
+                    new_v = dict(zip(keys, updated_values))
+                else:
+                    new_v = dict(zip(keys, values))
+            self.__setattr__(k, new_v)
+
+    def to_dict(self) -> Dict[str, ConfigsRecordValues]:
+        """Convert the state to a dictionary."""
+        ret = vars(self)
+        for k in list(ret.keys()):
+            if isinstance(ret[k], dict):
+                # Replace dict with two lists
+                v = cast(Dict[str, Any], ret.pop(k))
+                ret[f"{k}:K"] = list(v.keys())
+                if k == "public_keys_dict":
+                    v_list: List[bytes] = []
+                    for b1_b2 in cast(List[Tuple[bytes, bytes]], v.values()):
+                        v_list.extend(b1_b2)
+                    ret[f"{k}:V"] = v_list
+                else:
+                    ret[f"{k}:V"] = list(v.values())
+        return ret
+
+
+def _get_fit_fn(
+    msg: Message, call_next: FlowerCallable, ctxt: Context
+) -> Callable[[], FitRes]:
+    """Get the fit function."""
+
+    def fit() -> FitRes:
+        out_msg = call_next(msg, ctxt)
+        return compat.recordset_to_fitres(out_msg.message, keep_input=False)
+
+    return fit
+
+
+def secaggplus_middleware(
+    msg: Message,
+    ctxt: Context,
+    call_next: FlowerCallable,
+) -> Message:
+    """Handle incoming message and return results, following the SecAgg+ protocol."""
+    # Ignore non-fit messages
+    if msg.metadata.task_type != TASK_TYPE_FIT:
+        return call_next(msg, ctxt)
+
+    # Retrieve state
+    if RECORD_KEY_STATE not in ctxt.state.configs:
+        ctxt.state.set_configs(RECORD_KEY_STATE, ConfigsRecord({}))
```

Comment:
Should we do this check depending on the `current_stage`? If `current_stage` is not the initial stage, then having no `RECORD_KEY_STATE` key in `ctxt.state` would be an error, no?

Reply:
The current stage is stored in the `SecAggPlusState` as a `ConfigsRecord` in `Context.state`. We need to retrieve the current stage from `Context.state`. If no `RECORD_KEY_STATE` in the `ctxt.state`, we have to assume that the state is not initialised.

Do you mean the incoming stage we obtained from the server message? In this case, having no `RECORD_KEY_STATE` means the client is in the initial stage, and hence if the stage in the server message is not the first stage, an error will be raised.

Reply:
Yes, that's what I meant. If we received a server message with at stage that is not the initial stage, and there is not `RECORD_KEY_STATE` present, then an error should be raised because any stage that's not the initial stage should rely on the assumption that the initial stage took care of the state initialization (`ctxt.state.set_configs(RECORD_KEY_STATE, ConfigsRecord({}))`).